aiida/tools/importexport/archive/writers.py

```diff
@@ -35,176 +43,244 @@ def get_writer(file_format: str) -> Type['ArchiveWriterAbstract']:
+        return f'JSON Zip (compression={self._compression})'
 
     @property
     def export_version(self) -> str:
         return EXPORT_VERSION
 
-    def write(self, export_data: ArchiveData, get_node_repo: Callable[[str], Folder]) -> dict:
-        """write the archive
-
-        :param export_data: The data to export
-        :returns: A dictionary of data about the write process
-
-        """
-        with ZipFolder(self.filename, mode='w', use_compression=self._use_compression) as folder:
-            _write_to_json_archive(
-                folder=folder,
-                export_data=export_data,
-                export_version=self.export_version,
-                get_node_repo=get_node_repo,
-            )
-
-        return {}
+    def open(self):
+        # pylint: disable=attribute-defined-outside-init
+        self.assert_within_context()
+        # create a temporary folder in which to perform the write
+        self._temp_path: Path = Path(tempfile.mkdtemp())
+        # open a zipfile in in write mode to export to
+        self._zippath: ZipPath = ZipPath(self._temp_path / 'export', mode='w', compression=self._compression)
+        # setup data to store
+        self._data: Dict[str, Any] = {
+            'node_attributes': {},
+            'node_extras': {},
+            'export_data': {},
+            'links_uuid': [],
+            'groups_uuid': {},
+        }
+
+    def close(self, excepted: bool):
+        self.assert_within_context()
+        if excepted:
+            self._zippath.close()
+            shutil.rmtree(self._temp_path)
+            return
+        # write data.json
+        with self._zippath.joinpath('data.json').open('wb') as handle:
+            json.dump(self._data, handle)
+        # close the zipfile to finalise write
+        self._zippath.close()
+        # move the compressed file to the final path
+        self._remove_filepath()
+        shutil.move(str(self._zippath.filepath), str(self.filepath))
+        # remove the temporary folder
+        shutil.rmtree(self._temp_path)
+
+    def write_metadata(self, data: ArchiveMetadata):
+        metadata = {
+            'export_version': self.export_version,
+            'aiida_version': data.aiida_version,
+            'all_fields_info': data.all_fields_info,
+            'unique_identifiers': data.unique_identifiers,
+            'export_parameters': {
+                'graph_traversal_rules': data.graph_traversal_rules,
+                'entities_starting_set': data.entities_starting_set,
+                'include_comments': data.include_comments,
+                'include_logs': data.include_logs,
+            },
+            'conversion_info': data.conversion_info
+        }
+        with self._zippath.joinpath('metadata.json').open('wb') as handle:
+            json.dump(metadata, handle)
+
+    def write_link(self, data: Dict[str, str]):
+        self._data['links_uuid'].append(data)
```

Comment:
A long shot concerning the memory usage:
What exactly happens under the hood here?

Python lists are variable-length arrays, i.e. over the course of `write_link` being called N times for N links it probably needs to reallocate the array multiple times.

If the implementation is clever, it might realize that it needs to do this super often and therefore get more and more liberal in the size of the pre-allocation.

However, this goes to an implementation issue - at least for the current implementation, we know in advance how many links we will have, so it would make sense to communicate this to the writer in some way.

Reply:
this does not appear to be an issue (see performance analysis in comments)